src/libraries/Common/src/System/Reflection/Metadata/TypeName.cs

```diff
@@ -0,0 +1,329 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+
+namespace System.Reflection.Metadata
+{
+    [DebuggerDisplay("{AssemblyQualifiedName}")]
+#if SYSTEM_PRIVATE_CORELIB
+    internal
+#else
+    public
+#endif
+    sealed class TypeName : IEquatable<TypeName>
+    {
+        /// <summary>
+        /// Positive value is array rank.
+        /// Negative value is modifier encoded using constants defined in <see cref="TypeNameParserHelpers"/>.
+        /// </summary>
+        private readonly int _rankOrModifier;
+        private readonly TypeName[]? _genericArguments;
+        private readonly AssemblyName? _assemblyName;
+        private readonly TypeName? _elementOrGenericType;
+        private readonly TypeName? _declaringType;
+        private string? _assemblyQualifiedName;
+
+        internal TypeName(string name, string fullName,
+            AssemblyName? assemblyName,
+            int rankOrModifier = default,
+            TypeName? underlyingType = default,
+            TypeName? containingType = default,
+            TypeName[]? genericTypeArguments = default)
+        {
+            Name = name;
+            FullName = fullName;
```

Comment:
Precomputing the FullName makes the memory cost of the parser quadratic instead of linear.

My assumption was that the memory cost of the parser is going to be `O(size_of_input + max_nodes)`, but precomputing the FullName makes it `O(size_of_input * max_nodes)`.

Reply:
I assume that this is also one of the reasons why the new parser is much more allocation heavy. Calling `Type.GetType` on `typeof(Dictionary<List<int>,List<string>>).AssemblyQualifiedName` allocates 5kB before this change and 8.5kB with this change. It feels like a significant regression. What are the overall micro-benchmark results?

Maybe we should reconsider how all name properties on the TypeName should work, whether we need them all, and what their perf characteristics are.

Reply:
Do you agree that the memory cost of the parser should be `O(size_of_input + max_nodes)` (before you start calling any of the properties)?

Reply:
> compared to current parser

I agree it's a lot, but it's not a 100% apples-to-apples comparison (internal API doing one thing vs public API that allows for more and exposes more info).

I'll take a look at the code and see what can be made lazy-allocated.

Reply:
I've changed the implementation to allocate `Name` and `FullName` when actually needed. Here are the perf results:

The new implementation is always faster, but it may still allocate up to 2x more memory.

<details>

| Method                         | Job        | input                                | Mean        | Ratio | Allocated | Alloc Ratio |
|------------------------------- |----------- |------------------------------------- |------------:|------:|----------:|------------:|
| GetType_InvalidName            |         PR | ?                                    |   580.82 ns |  0.84 |         - |        0.00 |
| GetType_InvalidName            |       main | ?                                    |   688.86 ns |  1.00 |     112 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Collec(...)stem.Boolean] [69] | 3,627.66 ns |  0.79 |    1928 B |        0.88 |
| GetType_FullyQualifiedNames    |       main | System.Collec(...)stem.Boolean] [69] | 4,579.26 ns |  1.00 |    2184 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Collec(...)em.Boolean][] [71] | 3,731.70 ns |  0.80 |    2008 B |        0.91 |
| GetType_FullyQualifiedNames    |       main | System.Collec(...)em.Boolean][] [71] | 4,655.11 ns |  1.00 |    2216 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_NonFullyQualifiedNames |         PR | System.Int32                         |   925.25 ns |  0.99 |     128 B |        1.78 |
| GetType_NonFullyQualifiedNames |       main | System.Int32                         |   937.15 ns |  1.00 |      72 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Int32                         |   956.71 ns |  0.99 |     128 B |        1.78 |
| GetType_FullyQualifiedNames    |       main | System.Int32                         |   971.13 ns |  1.00 |      72 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_Resolvers              |         PR | System.Int32                         |    96.60 ns |  0.76 |     280 B |        1.25 |
| GetType_Resolvers              |       main | System.Int32                         |   127.63 ns |  1.00 |     224 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Int32&                        | 1,102.01 ns |  0.99 |     208 B |        2.00 |
| GetType_FullyQualifiedNames    |       main | System.Int32&                        | 1,116.70 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Int32*                        | 1,089.25 ns |  0.96 |     208 B |        2.00 |
| GetType_FullyQualifiedNames    |       main | System.Int32*                        | 1,134.69 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Int32[*]                      | 1,080.27 ns |  0.97 |     208 B |        2.00 |
| GetType_FullyQualifiedNames    |       main | System.Int32[*]                      | 1,117.69 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_NonFullyQualifiedNames |         PR | System.Int32[,]                      | 1,035.41 ns |  0.98 |     208 B |        2.00 |
| GetType_NonFullyQualifiedNames |       main | System.Int32[,]                      | 1,056.96 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Int32[,]                      | 1,091.25 ns |  0.95 |     208 B |        2.00 |
| GetType_FullyQualifiedNames    |       main | System.Int32[,]                      | 1,144.48 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_NonFullyQualifiedNames |         PR | System.Int32[]                       |   988.50 ns |  0.97 |     208 B |        2.00 |
| GetType_NonFullyQualifiedNames |       main | System.Int32[]                       | 1,019.59 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_NonFullyQualifiedNames |         PR | System.Int32[*]                      | 1,032.28 ns |  0.97 |     208 B |        2.00 |
| GetType_NonFullyQualifiedNames |       main | System.Int32[*]                      | 1,065.86 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_NonFullyQualifiedNames |         PR | System.Int32*                        |   998.85 ns |  0.98 |     208 B |        2.00 |
| GetType_NonFullyQualifiedNames |       main | System.Int32*                        | 1,022.85 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_NonFullyQualifiedNames |         PR | System.Int32&                        | 1,013.84 ns |  1.00 |     208 B |        2.00 |
| GetType_NonFullyQualifiedNames |       main | System.Int32&                        | 1,017.68 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Int32[]                       | 1,049.71 ns |  0.97 |     208 B |        2.00 |
| GetType_FullyQualifiedNames    |       main | System.Int32[]                       | 1,086.71 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_Resolvers              |         PR | System.Int32[]                       |   192.41 ns |  0.86 |     360 B |        1.41 |
| GetType_Resolvers              |       main | System.Int32[]                       |   224.20 ns |  1.00 |     256 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Tests.Perf_Type+Nested        | 1,129.14 ns |  0.98 |     456 B |        2.48 |
| GetType_FullyQualifiedNames    |       main | System.Tests.Perf_Type+Nested        | 1,150.65 ns |  1.00 |     184 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_Resolvers              |         PR | System.Tests.Perf_Type+Nested        |   198.32 ns |  0.89 |     608 B |        1.81 |
| GetType_Resolvers              |       main | System.Tests.Perf_Type+Nested        |   224.01 ns |  1.00 |     336 B |        1.00 |
|                                |            |                                      |             |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Tests.(...)stem.Boolean] [75] | 4,491.81 ns |  0.93 |    3864 B |        1.65 |
| GetType_FullyQualifiedNames    |       main | System.Tests.(...)stem.Boolean] [75] | 4,839.61 ns |  1.00 |    2336 B |        1.00 |

</details>

Reply:
Is the worst-case memory consumption of parsing `O(size_of_input + max_nodes)` now?

Reply:
I've optimized the code path for nested types and I believe it is now.


<details>

| Method                         | Job        | input                                                                       | Mean       | Ratio | Allocated | Alloc Ratio |
|------------------------------- |----------- |---------------------------------------------------------------------------- |-----------:|------:|----------:|------------:|
| GetType_InvalidName            |         PR | ?                                                                           |   578.0 ns |  0.86 |         - |        0.00 |
| GetType_InvalidName            |       main | ?                                                                           |   672.0 ns |  1.00 |     112 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Collections.Generic.Dictionary`2[System.String,System.Boolean]       | 3,595.9 ns |  0.79 |    1928 B |        0.88 |
| GetType_FullyQualifiedNames    |       main | System.Collections.Generic.Dictionary`2[System.String,System.Boolean]       | 4,535.1 ns |  1.00 |    2184 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Collections.Generic.Dictionary`2[System.String,System.Boolean][]     | 3,726.0 ns |  0.78 |    2008 B |        0.91 |
| GetType_FullyQualifiedNames    |       main | System.Collections.Generic.Dictionary`2[System.String,System.Boolean][]     | 4,792.3 ns |  1.00 |    2216 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_NonFullyQualifiedNames |         PR | System.Int32                                                                |   927.1 ns |  0.98 |     128 B |        1.78 |
| GetType_NonFullyQualifiedNames |       main | System.Int32                                                                |   945.8 ns |  1.00 |      72 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Int32                                                                |   953.2 ns |  0.95 |     128 B |        1.78 |
| GetType_FullyQualifiedNames    |       main | System.Int32                                                                |   999.3 ns |  1.00 |      72 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_Resolvers              |         PR | System.Int32                                                                |   101.9 ns |  0.80 |     280 B |        1.25 |
| GetType_Resolvers              |       main | System.Int32                                                                |   127.5 ns |  1.00 |     224 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Int32&                                                               | 1,081.1 ns |  0.96 |     208 B |        2.00 |
| GetType_FullyQualifiedNames    |       main | System.Int32&                                                               | 1,128.6 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Int32*                                                               | 1,088.2 ns |  0.95 |     208 B |        2.00 |
| GetType_FullyQualifiedNames    |       main | System.Int32*                                                               | 1,139.9 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Int32[*]                                                             | 1,083.2 ns |  0.97 |     208 B |        2.00 |
| GetType_FullyQualifiedNames    |       main | System.Int32[*]                                                             | 1,119.5 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_NonFullyQualifiedNames |         PR | System.Int32[,]                                                             | 1,023.6 ns |  0.97 |     208 B |        2.00 |
| GetType_NonFullyQualifiedNames |       main | System.Int32[,]                                                             | 1,056.7 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Int32[,]                                                             | 1,082.6 ns |  0.97 |     208 B |        2.00 |
| GetType_FullyQualifiedNames    |       main | System.Int32[,]                                                             | 1,118.0 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_NonFullyQualifiedNames |         PR | System.Int32[]                                                              |   995.7 ns |  0.97 |     208 B |        2.00 |
| GetType_NonFullyQualifiedNames |       main | System.Int32[]                                                              | 1,031.1 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_NonFullyQualifiedNames |         PR | System.Int32[*]                                                             | 1,033.3 ns |  0.99 |     208 B |        2.00 |
| GetType_NonFullyQualifiedNames |       main | System.Int32[*]                                                             | 1,048.0 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_NonFullyQualifiedNames |         PR | System.Int32*                                                               | 1,026.1 ns |  0.99 |     208 B |        2.00 |
| GetType_NonFullyQualifiedNames |       main | System.Int32*                                                               | 1,038.0 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_NonFullyQualifiedNames |         PR | System.Int32&                                                               |   998.5 ns |  0.98 |     208 B |        2.00 |
| GetType_NonFullyQualifiedNames |       main | System.Int32&                                                               | 1,019.8 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Int32[]                                                              | 1,056.5 ns |  0.97 |     208 B |        2.00 |
| GetType_FullyQualifiedNames    |       main | System.Int32[]                                                              | 1,085.6 ns |  1.00 |     104 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_Resolvers              |         PR | System.Int32[]                                                              |   194.5 ns |  0.83 |     360 B |        1.41 |
| GetType_Resolvers              |       main | System.Int32[]                                                              |   234.3 ns |  1.00 |     256 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Tests.Perf_Type+Nested                                               | 1,160.1 ns |  1.00 |     456 B |        2.48 |
| GetType_FullyQualifiedNames    |       main | System.Tests.Perf_Type+Nested                                               | 1,152.8 ns |  1.00 |     184 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_Resolvers              |         PR | System.Tests.Perf_Type+Nested                                               |   195.2 ns |  0.88 |     608 B |        1.81 |
| GetType_Resolvers              |       main | System.Tests.Perf_Type+Nested                                               |   221.3 ns |  1.00 |     336 B |        1.00 |
|                                |            |                                                                             |            |       |           |             |
| GetType_FullyQualifiedNames    |         PR | System.Tests.Perf_Type+Nested+NestedGeneric`2[System.String,System.Boolean] | 4,565.1 ns |  0.92 |    3784 B |        1.62 |
| GetType_FullyQualifiedNames    |       main | System.Tests.Perf_Type+Nested+NestedGeneric`2[System.String,System.Boolean] | 4,962.1 ns |  1.00 |    2336 B |        1.00 |

</details>

> Calling `Type.GetType` on `typeof(Dictionary<List<int>,List<string>>).AssemblyQualifiedName` 

This scenario is now 23% faster than main and allocates 12% less memory.